blog/2024-10-14-self-hosting-reflex-with-docker.md

```diff
@@ -0,0 +1,283 @@
+
+`WORKDIR /app` sets the working directory inside the docker container. `COPY . .` copies everything from your current directory into the docker container.
+
+The `ENTRYPOINT` is the command that is run when the container starts. In this case, we run the `reflex` command to start the Reflex app in production mode. The `--backend-only` flag tells Reflex to only run the backend server and not the frontend server. The `--loglevel` flag sets the log level to debug.
+
+
+### web.Dockerfile
+
+The `web.Dockerfile` is used to build the Docker image for the web server. Here is an example of a `web.Dockerfile`:
+
+```dockerfile
+FROM python:3.12 AS builder
+
+WORKDIR /app
+
+COPY . .
+RUN pip install -r requirements.txt
+RUN reflex export --frontend-only --no-zip
+
+FROM nginx
+
+COPY --from=builder /app/.web/_static /usr/share/nginx/html
+COPY ./nginx.conf /etc/nginx/conf.d/default.conf
+```
+
+We import `python:3.12` and build on top of it and the name of this build step is `builder` which we can then reference later as needed. The `WORKDIR`, `COPY`, `RUN pip install` commands are the same as in the `Dockerfile`. The `reflex export --frontend-only --no-zip` command exports the frontend assets to the `.web/_static` directory. After this step the build step we have with all these changes is called `builder`. 
+
+`nginx` is an image that we import and there is no need for an entry point as it already comes with an entry point built-in. 
+
+We now have `/app/.web/_static` in our builder from the reflex export command and we copy that into the location that nginx is looking to serve files from.
+
+We update the nginx configuration to one that serves our files and has a proxy pass to the backend as this is where nginx is looking to read a configuration file.
+
+### nginx.conf
+
+The `nginx.conf` file is used to configure the Nginx web server. Here is an example of an `nginx.conf` file:
+
+```nginx
+server { 
+ listen 80;
+ listen  [::]:80;
+ server_name frontend;
+
+
+ error_page   500 502 503 504  /50x.html;
+ location = /50x.html {
+    root   /usr/share/nginx/html;
+ }
+
+ location /_event {
+    proxy_set_header   Connection "upgrade";
+    proxy_pass http://backend:8000;
+    proxy_http_version 1.1;
+    proxy_set_header   Upgrade $http_upgrade;
+ }
+
+ location /ping {
+    proxy_pass http://backend:8000;
+ }
+
+ location /_upload {
+    proxy_pass http://backend:8000;
+ }
+
+ location / {
+   # This would be the directory where your Reflex app's static files are stored at
+   root /usr/share/nginx/html;
+   try_files $uri /$uri/index.html;
+ }
+
+}
+```
+
+The `listen 80;` and `listen  [::]:80;` lines specify that the server should listen on port 80 and on any ip address on port 80.
```

Comment:
[::] is for ipv6 not sure if we need to clarify that or not
```
IPv6 addresses (0.7.36) are specified in square brackets:

listen [::]:8000;
listen [::1];

```

but maybe remove `and on any ip address on port 80.` as it seems redunant?

Reply:
changed